src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNMethod.java

```diff
@@ -0,0 +1,271 @@
+    private boolean isCoarseQuantizerAvailable;
+
+    /**
+     * getMainMethodComponent
+     *
+     * @return mainMethodComponent
+     */
+    public MethodComponent getMethodComponent() {
+        return methodComponent;
+    }
+
+    /**
+     * Determines whether the provided space is supported for this method
+     *
+     * @param space to be checked
+     * @return true if the space is supported; false otherwise
+     */
+    public boolean hasSpace(SpaceType space) {
+        return spaces.contains(space);
+    }
+
+    /**
+     * Return the encoder with encoderName as name
+     *
+     * @param encoderName name of encoder to be looked up
+     * @return KNNEncoder corresponding to encoderName
+     */
+    public MethodComponent getEncoder(String encoderName) {
+        if (!encoders.containsKey(encoderName)) {
+            throw new IllegalArgumentException("Invalid encoder: " + encoderName);
+        }
+        return encoders.get(encoderName);
+    }
+
+    /**
+     * Determine if this method supports a given encoder
+     *
+     * @param encoderName name of encoder to be looked up
+     * @return true if
+     */
+    public boolean hasEncoder(String encoderName) {
+        return encoders.containsKey(encoderName);
+    }
+
+    /**
+     * isCoarseQuantizerAvailable
+     *
+     * @return true if coarse quantizer can be used with this method; false otherwise
+     */
+    public boolean isCoarseQuantizerAvailable() {
+        return isCoarseQuantizerAvailable;
+    }
+
+    /**
+     * Validate that the configured KNNMethodContext is valid for this method
+     *
+     * @param knnMethodContext to be validated
+     */
+    public void validate(KNNMethodContext knnMethodContext) {
+        if (!hasSpace(knnMethodContext.getSpaceType())) {
+            throw new ValidationException();
+        }
+
+        // validate the main method and its parameters
+        methodComponent.validate(knnMethodContext.getMethodComponent());
+
+        // validate the encoder and its parameters
+        KNNMethodContext.MethodComponentContext encoderContext = knnMethodContext.getEncoder();
+        if (encoderContext != null) {
+            try {
+                getEncoder(encoderContext.getName()).validate(encoderContext);
+            } catch (IllegalArgumentException iae) {
+                throw new ValidationException();
+            }
+        }
+    }
+
+    /**
+     * Generate extra parameters that do not go in the method string
+     *
+     * @param knnMethodContext from where parameters are retrieved
+     * @return map of extra parameters
+     */
+    public Map<String, Object> generateExtraParameterMap(KNNMethodContext knnMethodContext) {
+        return methodComponent.generateExtraParameterMap(knnMethodContext.getMethodComponent().getParameters());
+    }
+
+    /**
+     * MethodComponent defines the structure of an individual component that can make up an index
+     */
+    public static class MethodComponent {
```

Comment:
Have you consider using Builder pattern?

Reply:
Let me think about this. I think it could be used.

Reply:
Added it in the latest revision for KNNMethod and MethodComponent